# Pandas

In [3]:
# importa las bibliotecas pandas y numpy
import pandas as pd
import numpy as np

## Leer datos

Se utilizará la función `pandas.read_csv()` para leer el archivo *csv*.

Como los datos en el archivo de entrada no incluyen los encabezados, se puede agregar el argumento `headers = None` dentro del método `read_csv()` para que **pandas** no configure de forma automática que la primera fila corresponde al encabezado de los datos en el archivo de entrada.

El conjunto de datos se puede asignar a alguna variable para su creación.

In [5]:
# Lee el archivo de entrada y lo asigna a la variable "df"
ubicacion = "./datos/auto.csv"
df = pd.read_csv(ubicacion, header=None)

Después de leer el conjunto de datos, se puede empezar a utilizar el método `dataframe.head(n)` para verificar las primeras `n` filas del *dataframe*, donde `n` es un entero. Por otro lado, el método `dataframe.tail(n)` mostrará las últimas `n` filas del *dataframe*.


In [6]:
# muestra las primeras 5 filas usando el método dataframe.head()

print("Las primeras 5 filas del dataframe") 
df.head(5)

Las primeras 5 filas del dataframe


0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

### Agrega encabezados

Al observar el conjunto de datos leído se puede notar que **Pandas** automáticamente le asigna un encabezado con números enteros comenzando desde el 0.

Para una mejor descripción de los datos de entrada, se puede agregar un encabezado. Esta información está disponible en: [https://archive.ics.uci.edu/ml/datasets/Automobile](https://archive.ics.uci.edu/ml/datasets/Automobile).

Así, se podría agregar manualmente un encabezado.

Primero, se crea una lista "encabezado" que incluye todos los nombres de columnas en orden. Entonces, se utiliza `dataframe.columns = encabezado` para reemplazar los encabezados numéricos con la lista ya creada.

In [7]:
# crea una lista con el encabezado
encabezado = ["symboling","normalized-losses","make","fuel-type","aspiration","num-of-doors","body-style", "drive-wheels","engine-location","wheel-base","length","width","height","curb-weight","engine-type","num-of-cylinders","engine-size","fuel-system","bore","stroke","compression-ratio","horsepower","peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", encabezado)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


Se reemplaza el encabezado y se verifica que la tarea se realizó con éxito en el dataframe:

In [8]:
df.columns = encabezado
df.head(47)

symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           3                 ?  alfa-romero       gas        std   
2           1                 ?  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2                 ?         audi       gas        std   
6           1               158         audi       gas        std   
7           1                 ?         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0                 ?         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1                 ?          bmw       gas        std   
15          0                 ?          bmw       gas        std   
16          0                 ?          bmw       gas        std   
17          0                 ?          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas        std   
23          1               118        dodge       gas      turbo   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas        std   
27          1               148        dodge       gas      turbo   
28         -1               110        dodge       gas        std   
29          3               145        dodge       gas      turbo   
30          2               137        honda       gas        std   
31          2               137        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          1               101        honda       gas        std   
35          0               110        honda       gas        std   
36          0                78        honda       gas        std   
37          0               106        honda       gas        std   
38          0               106        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   
41          0                85        honda       gas        std   
42          1               107        honda       gas        std   
43          0                 ?        isuzu       gas        std   
44          1                 ?        isuzu       gas        std   
45          0                 ?        isuzu       gas        std   
46          2                 ?        isuzu       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105

Ahora se necesita reemplazar los signos de interrogación **?** con **NaN** así el método `dropna()` puede remover los valores perdidos.:


In [9]:
df1=df.replace('?',np.NaN)
df1.head(47)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0               NaN         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          0               NaN          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas        std   
23          1               118        dodge       gas      turbo   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas        std   
27          1               148        dodge       gas      turbo   
28         -1               110        dodge       gas        std   
29          3               145        dodge       gas      turbo   
30          2               137        honda       gas        std   
31          2               137        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          1               101        honda       gas        std   
35          0               110        honda       gas        std   
36          0                78        honda       gas        std   
37          0               106        honda       gas        std   
38          0               106        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   
41          0                85        honda       gas        std   
42          1               107        honda       gas        std   
43          0               NaN        isuzu       gas        std   
44          1               NaN        isuzu       gas        std   
45          0               NaN        isuzu       gas        std   
46          2               NaN        isuzu       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105

Así, ahora se pueden liberar los valores perdidos a lo largo de la columna **precio** como se muestra a continuación.


In [11]:
df=df1.dropna(subset=["price"], axis=0)
df.head(47)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          0               NaN          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas        std   
23          1               118        dodge       gas      turbo   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas        std   
27          1               148        dodge       gas      turbo   
28         -1               110        dodge       gas        std   
29          3               145        dodge       gas      turbo   
30          2               137        honda       gas        std   
31          2               137        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          1               101        honda       gas        std   
35          0               110        honda       gas        std   
36          0                78        honda       gas        std   
37          0               106        honda       gas        std   
38          0               106        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   
41          0                85        honda       gas        std   
42          1               107        honda       gas        std   
43          0               NaN        isuzu       gas        std   
46          2               NaN        isuzu       gas        std   
47          0               145       jaguar       gas        std   
48          0               NaN       jaguar       gas        std   
49          0               NaN       jaguar       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105

## Guardar un conjunto de datos
La biblioteca **Pandas** también provee la manera de salvar el conjunto de datos en un archivo con formato **csv**. Utilizando el método `dataframe.to_csv()`.

Por ejemplo, si se quisiera guardar el dataframe `df` como `automoviles.csv` a un directorio de la máquina local, Se debe usar la sintaxis mostrada a continuación, donde `index = False` significa que el nombre de las columnas no será escrito.

In [8]:
df.to_csv("./datos/automoviles.csv", index=False)

# Información básica de los conjuntos de datos

Después de leer los datos en un dataframe **Pandas**, es tiempo de explorar el conjunto de datos.


## Tipos de datos

Los datos tienen una variedad de tipos.

El principal tipo de datos almacenados en una dataframe de **Pandas** son **object**, **float**, **int**, **bool** y **datetime64**. Para entender mejor acerca de cada atributo, es siempre bueno conocer el tipo de datos de cada columna. En **Pandas** se utiliza el atributo `dtypes`.


Así, una serie con los tipos de datos de cada columna es retornado.


In [9]:
# Verifica el tipo de datos del dataframe "df" usando .dtypes
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Como se muestra arriba, se puede ver claramente que el tipo de datos de las columnas **symboling** y **curb-weight** son `int64`, **normalized-losses** es `object`, y **wheel-base** es `float64`, entre otros.




## Descripción

Si quisiéramos tener un resumen estadístico de cada columna, por ejemplo, la cantidad, el promedio de la columna la desviación estándar de la columna, entre otros, se debe utilizar el método `.describe()`.


Este método proveerá un resumen de varios datos estadísticos de cada columna, excluyendo los valores `NaN` (Not a Number).


In [10]:
df.describe()

symboling  wheel-base      length       width      height  \
count  201.000000  201.000000  201.000000  201.000000  201.000000   
mean     0.840796   98.797015  174.200995   65.889055   53.766667   
std      1.254802    6.066366   12.322175    2.101471    2.447822   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.800000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.500000   66.600000   55.500000   
max      3.000000  120.900000  208.100000   72.000000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   201.000000   201.000000         201.000000  201.000000   201.000000  
mean   2555.666667   126.875622          10.164279   25.179104    30.686567  
std     517.296727    41.546834           4.004965    6.423220     6.815150  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2169.000000    98.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2926.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

Esto muestra el resumen estadístico de toda las columnas que son de tipo numérico (int, float).

Por ejemplo, el atributo **symboling** tiene 205 elementos, el valor promedio de esta columna es 0.83, la desviación estándar es 1.25, El valor mínimo es -2, el percentil del 25% es 0, el percentil del 50% es 1, el percentil del 75% es 2, y el valor máximo es 3.

Sin embargo, que pasa si se quiere chequear todas las columnas, incluyendo esas que son del tipo `object`.

Se puede agregar un argumento `include = "all"` dentro de los paréntesis. Probemos.


In [11]:
# descrive todas las columnas en "df" 
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   201.000000               164     201       201        201   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       181        165   
mean      0.840796               NaN     NaN       NaN        NaN   
std       1.254802               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           199        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    201.000000          201   197     197        201.000000        199   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           92    23      19               NaN         19   
mean     126.875622          NaN   NaN     NaN         10.164279        NaN   
std       41.546834          NaN   NaN     NaN          4.004965        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       98.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        199  201.000000  201.000000   201  
unique        22         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          36         NaN         NaN     2  
mean         NaN   25.179104   30.686567   NaN  
std          NaN    6.423220    6.815150   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

Aquí, se muestra el resumen estadístico de todas las columnas, incluyendo las columnas que son del tipo `object`.

Así, se puede observar cuantos valores hay que son únicos, Cual valor es el top y la frecuencia de ese valor top en las columnas que son del tipo `object`.

Algunos valores en la tabla de arriva se muestran como **NaN**. Esto es porque esos números no están disponiblescon respecto al tipo particular de alguna columna.

## Información

Otro método que se puede utilizar para verificar el conjunto de datos es, `dataframe.info()`.


Este provee un resumen consiso del DataFrame.

Este método imprime información acerca de un DataFrame incluyendo los índices `dtype` y columnas, valores no-nulos y uso de la memoria.


In [12]:
# muestra la información del "df"

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non